### PR TITLE
feat(router): overhaul route discovery engine

### LINKED ISSUE #237: RFC: Add support for pathless layout routes
### Which scope/s are relevant/related to the feature request?

router

### Information

This feature can be useful to group related pages into the same folder without adding additional segments to the URL.

For the following structure:

```
app
  /routes
    /(auth)
       login.ts
       signup.ts
```

we'll be able to access the `login` and `signup` pages by using `/login` and `/signup` URLs. In the same way, we should be able to define a layout/shell component for pages in the `(auth)` folder:

```
app
  /routes
    /(auth)
       login.ts
       signup.ts
    (auth).ts
```

---

Pathless layout routes in Remix: https://remix.run/docs/en/v1/guides/routing#pathless-layout-routes

### Describe any alternatives/workarounds you're currently using

_No response_

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by markostanimirovic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/analogjs/analog/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [ ] vite-plugin-angular
- [ ] vite-plugin-nitro
- [ ] astro-angular
- [ ] create-analog
- [x] router
- [ ] platform
- [ ] content
- [ ] nx-plugin
- [ ] trpc

## What is the current behavior?

- It's not possible to create named nested pages without a parent (layout) page.
- Creating pages within the pathless directory and pathless layouts are not supported.

Closes #237, #273

## What is the new behavior?

- We can create named nested routes without a layout route.

```
/pages
  /blog
    posts.page.ts
```

- We can create pages within the pathless directories.

```
/pages
  /(auth)
    login.page.ts
    signup.page.ts
```

- We can define the layout page for pages within the pathless directory.

```
/pages
  /(auth)
    login.page.ts
    signup.page.ts
  (auth).page.ts
```

- We can create nested wildcard routes:

```
/pages
  /blog
    posts.page.ts
    [...blog-not-found].md
  blog.page.ts
  index.page.ts
  [...not-found].md
```

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## [optional] What [gif](https://chrome.google.com/webstore/detail/gifs-for-github/dkgjnpbipbdaoaadbdhpiokaemhlphep) best describes this PR or how it makes you feel?

hooray-letsgo

**EDIT:** 😅 

magic

### CODE REVIEW

File: packages/router/src/lib/routes.ts [hunk 1]
```diff
@@ -14,145 +14,169 @@ const FILES = import.meta.glob<RouteExport>([
...
-                ]),
-            });
-          }
-
-          return (
-            found ||
-            (current?.[insert === 'unshift' ? 0 : current.length - 1] as Route)
-          );
-        }
+  const rootRawRoutesMap = rawRoutesByLevelMap[0] || {};
```
Comment (brandonroberts) [inline packages/router/src/lib/routes.ts:92]: Under what scenario would this array be empty? If no routes were found?

**Review by brandonroberts — APPROVED** [review]

File: packages/router/src/lib/routes.ts (hunk 1 above)
Comment (markostanimirovic) [reply packages/router/src/lib/routes.ts:92]: Yes, if `files` is an empty object. I can make this check more explicit and move it to the top of the method:

```ts
function createRoutes(files: Files): Route[] {
  const filenames = Object.keys(files);

  if (filenames.length === 0) {
    return [];
  }

  // ...
}
```

what do you think? 👀

Comment (brandonroberts) [reply packages/router/src/lib/routes.ts:92]: Works for me 👍

**Comment by zawasp** [comment]
Does this by any chance also fix #412 ?

**Comment by matheo** [comment]
@brandonroberts it would be nice to have a command printing the detected routes somehow,
I tried to check the generated code with no luck

**Comment by brandonroberts** [comment]
@matheo routing nested in folders is supported. What do you mean by using dot notation inside nested folders? Any file with a .page.ts extension should be picked up.

You can log the routes

import { routes } from '@analogjs/router';

console.log(routes);

**Pull request merged**